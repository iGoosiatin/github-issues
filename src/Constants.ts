export const ITEMS_PER_PAGE = 10;

export const SCREENS = {
  HOME: 'Home',
  ISSUES: 'Issue',
};
