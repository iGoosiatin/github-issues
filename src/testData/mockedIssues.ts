export const issues = [
  {
    id: 1325995261,
    number: 1297,
    title: '[Question] Extra context in formsets - additional examples / explanation for rendering field values as text in formsets?',
    user: {
      login: 'rzrwolf',
    },
    comments: 0,
    created_at: '2022-08-02T15:05:29Z',
  },
  {
    id: 1309663812,
    number: 1296,
    title: 'Feature Request: configurable global input size instead of bootstrap default size',
    user: {
      login: 'NameX44',
    },
    comments: 0,
    created_at: '2022-07-19T15:05:37Z',
  },
  {
    id: 1299872111,
    number: 1294,
    title: '2.0a1 release',
    user: {
      login: 'smithdc1',
    },
    comments: 2,
    created_at: '2022-07-10T09:22:06Z',
    pull_request: {},
  },
];
