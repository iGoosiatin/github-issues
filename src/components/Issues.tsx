import React, { useContext, useEffect, useState, useReducer, useCallback } from 'react';
import { ActivityIndicator, Text, StyleSheet, View } from 'react-native';
import { loadIssues } from '../services/loadIssues';
import { SearchDataContext } from '../Context';
import { MemoizedSort } from './Sort';
import IssueList from './IssueList';
import { MemoizedPaginationControls } from './Pagination/PaginationControls';
import issueReducer, { IIssueReducerState } from '../reducers/issueReducer';
import { ITEMS_PER_PAGE } from '../Constants';
import { sortOptions } from '../Utils';

import ISearchData from '../types/SearchData';
import { ILoadIssuesResponseFailure, ILoadIssuesResponseSuccess } from '../types/LoadIssuesReponse';
import { SortByType, SortDirectionType } from '../types/Sort';

const initialState: IIssueReducerState = {
  issues: [],
  sortBy: 'created',
  sortDirection: 'desc',
  selectedSortOption: 0,
  currentPage: 1,
  previousPage: null,
  nextPage: null,
  lastPage: 1,
};

const Issues = () => {
  const { searchData } = useContext(SearchDataContext);

  const [isLoading, setIsLoading] = useState(false);
  const [errorText, setErrorText] = useState<String>('');

  const [state, dispatch] = useReducer(issueReducer, initialState);

  const setPage = useCallback((page: number | null) => {
    if (page === null) {
      return;
    }
    dispatch({ type: 'SET_PAGE', payload: page });
  }, []);

  const onSort = useCallback((selectedIndex: number) => {
    dispatch({ type: 'SET_SORTING_OPTION', payload: selectedIndex });
  }, []);

  useEffect(() => {
    const handleLoadIssues = async (
      { user, repo }: ISearchData,
      page: number,
      sortBy: SortByType,
      sortDirection: SortDirectionType,
    ) => {
      setIsLoading(true);
      const response = await loadIssues({ user, repo, page, sortBy, sortDirection });
      if (response.isError) {
        const failedLoad = response as ILoadIssuesResponseFailure;
        setErrorText(failedLoad.errorText);
        dispatch({ type: 'SET_ISSUES', payload: { issues: [], lastPage: 1 } });
      } else {
        const successLoad = response as ILoadIssuesResponseSuccess;
        setErrorText('');
        dispatch({
          type: 'SET_ISSUES',
          payload: {
            issues: successLoad.issues,
            lastPage: Math.ceil(successLoad.openIssues / ITEMS_PER_PAGE),
          },
        });
      }
      setIsLoading(false);
    };
    if (searchData !== null) {
      handleLoadIssues(searchData, state.currentPage, state.sortBy, state.sortDirection);
    }
  }, [searchData, state.currentPage, state.sortBy, state.sortDirection]);

  if (isLoading) {
    return (
      <View style={styles.loading}>
        <ActivityIndicator size="large" color="gray" />
      </View>
    );
  }

  if (errorText) {
    return <Text style={styles.errorText}>{errorText}</Text>;
  }

  return (
    <View style={styles.container}>
      <View>
        <MemoizedSort options={sortOptions} selectedIndex={state.selectedSortOption} onSort={onSort} />
      </View>
      <View style={styles.list}>
        <IssueList issues={state.issues} />
      </View>
      <MemoizedPaginationControls
        currentPage={state.currentPage}
        previousPage={state.previousPage}
        nextPage={state.nextPage}
        lastPage={state.lastPage}
        setPage={setPage}
      />
    </View>
  );
};

const styles = StyleSheet.create({
  loading: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
  },
  errorText: {
    marginTop: 10,
    textAlign: 'center',
    color: 'red',
  },
  container: {
    flex: 1,
  },
  list: {
    flex: 1,
  },
});

export default Issues;
