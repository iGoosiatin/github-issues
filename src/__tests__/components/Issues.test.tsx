import React from 'react';
import Issues from '../../components/Issues';
import { issues } from '../../testData/mockedIssues';
import { ITEMS_PER_PAGE } from '../../Constants';

import { render, waitFor } from '@testing-library/react-native';

describe('Issue component', () => {
  // Mocked fetch to return issues
  (globalThis.fetch as jest.Mock) = jest
    .fn(
      (): Promise<any> =>
        Promise.resolve({
          ok: true,
          json: () => Promise.resolve(issues),
        }),
    )
    .mockImplementationOnce(
      (): Promise<any> =>
        Promise.resolve({
          ok: true,
          // Should give 2 fake pages
          json: () => Promise.resolve({ open_issues_count: ITEMS_PER_PAGE + 1 }),
        }),
    )
    .mockImplementationOnce(
      (): Promise<any> =>
        Promise.resolve({
          ok: true,
          json: () => Promise.resolve(issues),
        }),
    )
    .mockImplementationOnce(
      (): Promise<any> =>
        Promise.resolve({
          ok: true,
          // Should give 2 fake pages
          json: () => Promise.resolve({ open_issues_count: ITEMS_PER_PAGE + 1 }),
        }),
    );
  it('Renders correctly', async () => {
    const screen = await waitFor(() => render(<Issues searchData={{ user: 'user', repo: 'repo' }} />));
    expect(screen).toMatchSnapshot();
  });
});
