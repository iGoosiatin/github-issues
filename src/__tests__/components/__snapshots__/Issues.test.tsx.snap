// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Issue component Renders correctly 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "borderBottomColor": "darkgray",
        "borderBottomWidth": 2,
        "flexDirection": "row",
        "marginTop": 5,
        "paddingBottom": 5,
      }
    }
  >
    <View
      style={
        Object {
          "marginLeft": 5,
          "width": "30%",
        }
      }
    >
      <Text>
        Sort By:
      </Text>
      <View
        onLayout={[Function]}
        style={
          Array [
            Object {
              "justifyContent": "center",
            },
            Object {
              "backgroundColor": "#FFF",
              "borderColor": "#444",
              "borderWidth": 1,
              "height": 30,
              "paddingLeft": 5,
              "width": "100%",
            },
          ]
        }
      >
        <View
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignItems": "center",
              "flexDirection": "row",
              "height": 35,
              "justifyContent": "space-between",
            }
          }
        >
          <Text
            style={
              Array [
                Object {
                  "flex": 1,
                  "fontSize": 16,
                },
                Object {
                  "fontSize": 12,
                },
                Object {},
              ]
            }
          >
            Newest
          </Text>
          <Image
            source={
              Object {
                "testUri": "../../react-native-element-dropdown/lib/commonjs/assets/down.png",
              }
            }
            style={
              Array [
                Object {
                  "height": 20,
                  "width": 20,
                },
                Object {
                  "tintColor": "gray",
                },
                undefined,
              ]
            }
          />
        </View>
      </View>
    </View>
    <View
      style={
        Object {
          "marginLeft": 5,
          "width": "30%",
        }
      }
    >
      <Text>
        Filter By:
      </Text>
      <View
        onLayout={[Function]}
        style={
          Array [
            Object {
              "justifyContent": "center",
            },
            Object {
              "backgroundColor": "#FFF",
              "borderColor": "#444",
              "borderWidth": 1,
              "height": 30,
              "paddingLeft": 5,
              "width": "100%",
            },
          ]
        }
      >
        <View
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignItems": "center",
              "flexDirection": "row",
              "height": 35,
              "justifyContent": "space-between",
            }
          }
        >
          <Text
            style={
              Array [
                Object {
                  "flex": 1,
                  "fontSize": 16,
                },
                Object {
                  "fontSize": 12,
                },
                Object {},
              ]
            }
          >
            Select filter
          </Text>
          <Image
            source={
              Object {
                "testUri": "../../react-native-element-dropdown/lib/commonjs/assets/down.png",
              }
            }
            style={
              Array [
                Object {
                  "height": 20,
                  "width": 20,
                },
                Object {
                  "tintColor": "gray",
                },
                undefined,
              ]
            }
          />
        </View>
      </View>
    </View>
  </View>
  <View
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <RCTScrollView
      data={
        Array [
          Object {
            "comments": 0,
            "created_at": "2022-08-02T15:05:29Z",
            "id": 1325995261,
            "number": 1297,
            "title": "[Question] Extra context in formsets - additional examples / explanation for rendering field values as text in formsets?",
            "user": Object {
              "login": "rzrwolf",
            },
          },
          Object {
            "comments": 0,
            "created_at": "2022-07-19T15:05:37Z",
            "id": 1309663812,
            "number": 1296,
            "title": "Feature Request: configurable global input size instead of bootstrap default size",
            "user": Object {
              "login": "NameX44",
            },
          },
          Object {
            "comments": 2,
            "created_at": "2022-07-10T09:22:06Z",
            "id": 1299872111,
            "number": 1294,
            "pull_request": Object {},
            "title": "2.0a1 release",
            "user": Object {
              "login": "smithdc1",
            },
          },
        ]
      }
      getItem={[Function]}
      getItemCount={[Function]}
      keyExtractor={[Function]}
      onContentSizeChange={[Function]}
      onLayout={[Function]}
      onMomentumScrollBegin={[Function]}
      onMomentumScrollEnd={[Function]}
      onScroll={[Function]}
      onScrollBeginDrag={[Function]}
      onScrollEndDrag={[Function]}
      removeClippedSubviews={false}
      renderItem={[Function]}
      scrollEventThrottle={50}
      stickyHeaderIndices={Array []}
      viewabilityConfigCallbackPairs={Array []}
    >
      <View>
        <View
          onLayout={[Function]}
          style={null}
        >
          <View
            style={
              Object {
                "borderBottomColor": "darkgray",
                "borderBottomWidth": 2,
                "padding": 5,
              }
            }
          >
            <Text
              style={
                Object {
                  "fontWeight": "bold",
                }
              }
            >
              [Issue] 
              [Question] Extra context in formsets - additional examples / explanation for rendering field values as text in formsets?
            </Text>
            <Text>
              #1297 opened on 02 Aug 2022 by rzrwolf
            </Text>
          </View>
        </View>
        <View
          onLayout={[Function]}
          style={null}
        >
          <View
            style={
              Object {
                "borderBottomColor": "darkgray",
                "borderBottomWidth": 2,
                "padding": 5,
              }
            }
          >
            <Text
              style={
                Object {
                  "fontWeight": "bold",
                }
              }
            >
              [Issue] 
              Feature Request: configurable global input size instead of bootstrap default size
            </Text>
            <Text>
              #1296 opened on 19 Jul 2022 by NameX44
            </Text>
          </View>
        </View>
        <View
          onLayout={[Function]}
          style={null}
        >
          <View
            style={
              Object {
                "borderBottomColor": "darkgray",
                "borderBottomWidth": 2,
                "padding": 5,
              }
            }
          >
            <Text
              style={
                Object {
                  "fontWeight": "bold",
                }
              }
            >
              [PR] 
              2.0a1 release
            </Text>
            <Text>
              #1294 opened on 10 Jul 2022 by smithdc1
            </Text>
            <Text>
              2 comment(s)
            </Text>
          </View>
        </View>
      </View>
    </RCTScrollView>
  </View>
  <View
    style={
      Object {
        "flexDirection": "row",
        "justifyContent": "center",
        "margin": 10,
      }
    }
  >
    <View
      accessibilityState={
        Object {
          "disabled": true,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
    >
      <Text
        style={
          Array [
            Object {
              "padding": 5,
            },
            false,
            Object {
              "color": "lightgray",
            },
          ]
        }
      >
        &lt;
      </Text>
    </View>
    <View
      accessibilityState={
        Object {
          "disabled": true,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "backgroundColor": "darkgray",
          "opacity": 1,
        }
      }
    >
      <Text
        style={
          Array [
            Object {
              "padding": 5,
            },
            Object {
              "color": "white",
              "fontWeight": "bold",
            },
            false,
          ]
        }
      >
        1
      </Text>
    </View>
    <View
      accessibilityState={
        Object {
          "disabled": false,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
    >
      <Text
        style={
          Array [
            Object {
              "padding": 5,
            },
            false,
            false,
          ]
        }
      >
        2
      </Text>
    </View>
    <View
      accessibilityState={
        Object {
          "disabled": false,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
    >
      <Text
        style={
          Array [
            Object {
              "padding": 5,
            },
            false,
            false,
          ]
        }
      >
        &gt;
      </Text>
    </View>
  </View>
</View>
`;
